lib/pbench/server/api/resources/datasets_list.py

```diff
@@ -294,61 +294,65 @@ def filter_query(filters: list[str], query: Query) -> Query:
 
         return query.filter(and_(*and_list))
 
-    def accumulate(self, keys: JSONOBJECT, key: str, value: Any):
+    def accumulate(self, aggregate: JSONOBJECT, key: str, value: Any):
         """Recursive helper to accumulate the metadata namespace
 
         Iterate through a list of metadata key/value pairs to construct a
-        herarchical aggregation of all metadata keys across the selected
+        hierarchical aggregation of all metadata keys across the selected
         datasets. Each key in the hierarchy is represented as a key in a
-        nested JSON object. "Leaf" keys are represented with empty JSON
-        values. E.g.,
+        nested JSON object. "Leaf" keys have the value None. E.g.,
 
             {
-                "dataset": {"name": {}, "metalog": {"pbench": {"script": {}}}},
-                "server": {"deletion": {}, "tarball-path": {}},
-                "global": {"server": {"legacy": {"sha1": {}}}}
+                "dataset": {"name": None, "metalog": {"pbench": {"script": None}}},
+                "server": {"deletion": None, "tarball-path": None},
+                "global": {"server": {"legacy": {"sha1": None}}}
             }
 
         Args:
-            keys: a JSONOBJECT to update with the recursive key/value
+            aggregate: a JSONOBJECT to update with the recursive key/value
             key: the current metadata key path element
             value: the current metadata key's value
         """
-        if key in keys:
-            p = keys[key]
-        else:
-            p = {}
-            keys[key] = p
         if isinstance(value, dict):
+            p = aggregate.get(key)
+            if p is None:
+                p = {}
+                aggregate[key] = p
             for k, v in value.items():
                 self.accumulate(p, k, v)
+        elif key not in aggregate:
+            aggregate[key] = None
```

Comment:
Setting the value of leaf nodes to `None` instead of `{}` adds some awkwardness to the code and I think it creates an opportunity for a conflict (which your code here papers over...perhaps on purpose...I think the result is the same as your previous code).

Unfortunately, we have to ask two questions, now -- is the `key` already present in the `aggregate`, and is the `value` a dictionary -- and the answer to one doesn't save us from having to ask the other.

So, here's an alternate casting of the above code:
```
        if key not in aggregate:
            aggregate[key] =  {} if isinstance(value, dict) else None
        if isinstance(value, dict):
            assert isinstance(aggregate[key], dict)
            for k, v in value.items():
                self.accumulate(aggregate[key], k, v)
```
(If we ask the questions in the other order, we get code very like what you proposed.)

Note that I'm not actually suggesting introducing an assertion; however, that's the point at which the conflict would be detected -- this would be a place where one dataset had a value for a key where another dataset has a sub-space, and I'm not sure how this should be represented in the resulting aggregate.  (In your proposed code, you replace the existing `None` entry with a subspace...that's not wrong, but I'm concerned that it will make it hard for a user to know that in some cases there is a value there.)  We should probably do something like place a list at this node in the tree (or add some special key to the dictionary), but neither of those is very "good".

Reply:
`None` was what Varshini wanted. It wasn't that big a deal. There is a potential for "conflict" across datasets because a leaf node in some (particularly under `user` or `global`) might be a sub-tree under others. The metadata model is that any partial path is legal: if you have `global.server.legacy.sha1`, you're allowed to get `global.server.legacy.sha1`, `global.server.legacy` (a JSON document of `{"sha1": <value>}`, `global.server` (a JSON document of `{"legacy": {"sha1": <value}}`, or just `global` to get everything.

So if `global.server` is a leaf in one dataset and a subdocument in another, that isn't really an issue in representing or using the namespace. We might have hit `global.server` first and set it to `None` and then found `global.server.legacy` and changed it to `{"legacy": {...}}` and that's not a logical problem in this API. (And if we encounter them in the other order, the key already exists and it's left alone.)

One nuance relates to the warning message you pointed out. If one dataset has `"global": {"server": "mine"}` and another has `"global": {"server": {"legacy": {"sha1": "bigstring"}}}` and you ask for ?metadata=global.server.legacy.sha1, you'll get that value for the latter case and no output (with the logged warning message) for the former case because there *is* no `global.server.legacy.sha1` and can't be. There's no practical (or probably useful) way to represent this in the aggregated output, though: it's just the nature of the beast. *But* I could in theory report the error, either locally like this or maybe globally (e.g., I could put the `errors` list on the outside of the `results` object in the `/datasets` response, aggregating them across all datasets).

```
{
    "name": <name>,
    "resource_id": <id>,
    "metadata": {...}, # whatever else
    "errors": ["Key 'global.server' value for 'global.server.legacy.sha1' in <dataset> is not a JSON object"]
}
```

I'm just not sure I want to do that in this PR as it's really got nothing to do with the aggregation.

Reply:
> if `global.server` is a leaf in one dataset and a subdocument in another, that isn't really an issue in representing or using the namespace.

The problem crops up on the client side:  we have to be clear that _any_ node in the returned "tree" is a valid key for obtaining metadata, but that the client might find a scalar value there or might find a JSON object (or might find `None`).

I'm expecting that the UI's principal use for this will be to allow the user to find scalar values, which means that if the aggregation doesn't annotate "conflict" nodes in some way, it will be awkward for the UI to help the user to find datasets with scalar values at those locations.  (That is, if some of the datasets in the aggregation have a value at `global.server.legacy.sha1`, I'm concerned that that will make it challenging for a user to discover that other datasets have scalar values at `global.server.legacy`.)

I wonder if it would be helpful to provide a count of scalar values at each node.  That would make it clear when there are values in the middle of the tree, and it might also allow the client to prioritize "popular" keys.

Reply:
> There's no practical (or probably useful) way to represent this in the aggregated output, though

I wouldn't go that far.  (We could do [counts](https://github.com/distributed-system-analysis/pbench/pull/3345#discussion_r1137412079); we could provide a list of the [flattened keys](https://github.com/distributed-system-analysis/pbench/pull/3345#pullrequestreview-1340252395) which have values; we could use fancier node representations, et al.)

> I could in theory report the error, either locally like this or maybe globally (e.g., I could put the errors list on the outside of the results object in the /datasets response, aggregating them across all datasets).

For a bulk request (e.g., "give me all of the `global.server.legacy.sha1` values for this list of datasets"), I think it's satisfactory to just return `None` for any datasets which don't have that key defined.  I don't _think_ we need to report any errors in the cases where the individual datasets don't have the value.  (And, in any case, I think that's outside the scope of this PR.)

Reply:
A count of matches sounds tempting, but I'm not crazy about the representation. In order to handle nodes which, across the set of datasets, are both leaf and non-leaf, I'd need to make each node an object with a sub-tree link and a count. Besides, if we're looking for "popularity" metrics, non-leaf counts might be useful too. Everyone has a `dataset`; most everyone has a `server`, many might not have `user` or `global` at all; right now all pass-through datasets will have `global.server.legacy`, whereas it's unlikely that any direct `PUT` datasets would ... etc.

But we end up with

```
{
    "dataset": {
        {"id": {"count": 10}}
        {"metalog": {
            "count": 10,
            "children": {
                "pbench": {
                    "count": 10,
                    "children": {
                        "script": {"count": 10}
            }
    },
    "global": {
        "count": 5,
        "children": {
            "server": {
                "count": 2,
                "children": {
                    "server": {...}
                }
            },
           "dashboard": {
               "count": 3,
               "children": {
                   "seen": {"count": 2},
                   "saved": {"count": 3}
               }
           }
        }
    }
}
```

Or would we want "count-as-parent" and "count-as-scalar" separately? I'm not sure I like where this would lead, though it might have value.

But as for the "to find scalar values", I'm not so sure about that. Metadata is JSON-based, and I can see a lot of value in aggregates like Peter's `global.server`. Sure, you can select display of `global.server.legacy.sha1` and `global.server.legacy.version` and `global.server.legacy.host`, but I'd prefer to be able to have the UI pretty-print the `global.server` JSON instead.

Reply:
> I'd prefer to be able to have the UI pretty-print the global.server JSON instead.

I have my doubts as to whether JSON _can_ be pretty-printed, but...that aside, I suspect whether we want to see the JSON or the individual values will vary depending on the key in question.  🦐 

Regardless, I'm expecting that the results of this aggregation are going to be used for one of two things:  to select particular datasets for display in some table, and to select particular values in the data (e.g., columns) to be displayed in the rows of it.  In that realm, JSON blobs are going to be awkward to deal with:  doing "equal to", "greater than", "contains", et al. operations which are intended to target a scalar should presumably return `False` when applied to a JSON object, which means that users will want/need/etc. to be able to find out which keys will hit scalars before running the selection; and the resulting display probably won't (or shouldn't) have much vertical space (by default, at least) for displaying JSON blobs, regardless of how we prettify them (otherwise, there won't be room for very many rows on the screen).

So, I'm looking for ways to make it easy for the UI to identify keys which will yield scalar values for the user (hence having the Server return a list of flattened keys instead of the object-hierarchy -- we could return both `global.server.legacy` and `global.server.legacy.sha1` if there are scalars to be found at both of them in different datasets).  Annotating the nodes with counts is a means toward allowing the UI to find those keys itself...but, as you show, it's cumbersome in terms of the response format and it will force the UI to walk the tree as well.

But, maybe my conception of the user's and UI's needs is misplaced.